src/client/connect/dns.rs

```diff
@@ -245,6 +245,56 @@ impl Iterator for IpAddrs {
     }
 }
 
+#[derive(Debug)]
+pub(super) struct Ipv6AddrWithZone {
+    pub(super) partial_addr: SocketAddrV6,
+    pub(super) as_str: String,
+}
+
+impl Ipv6AddrWithZone {
+    pub(super) fn try_parse(host: &str, port: u16) -> Option<Self> {
+        // IpV6 literals are always in []
+        if ! host.starts_with("[") {
+            return None;
+        }
+
+        let host = host.trim_start_matches('[').trim_end_matches(']');
```

Comment:
This will ignore `[[fe80::1:2:3:4%25eth0]`, which isn't allowed as far as I can tell. Instead, I'd recommend:

```
if !host.starts_with('[') || !host.ends_with(']') {
    return None;
}
let host = &host[1..host.len()-1];
```

Also, there should be a test case for erring out if the brackets are incorrect.

Reply:
Note that this mirrors the IPv6 literal parsing, above.

Reply:
true, but I think that'd be a bug there too :)